syntax = "proto3";

package auth;

import "google/protobuf/timestamp.proto";
import "pb_common/result.proto";

// ==================== 枚举 ====================

enum VerifyCodeType {
    VERIFY_CODE_TYPE_UNSPECIFIED = 0;
    REGISTER = 1;
    LOGIN = 2;
    RESET_PASSWORD = 3;
}

// ==================== 通用消息 ====================

// 用户信息（登录/注册返回用）
message UserInfo {
    string id = 1;                                  // uuid
    string mobile = 2;                              // 手机号
    string display_name = 3;                        // 用户名
    bool disabled = 4;                              // 是否被禁用
    google.protobuf.Timestamp created_at = 5;       // 创建时间
}

// Token 对
message TokenPair {
    string access_token = 1;
    string refresh_token = 2;
    //expires_in 指 Access Token 的过期时间，因为客户端只需要关心 Access Token 何时过期。——> 客户端可能需要在过期前主动刷新  
    int64 expires_in = 3;
}

// ==================== 验证码 ====================

message SendVerifyCodeRequest {
    string mobile = 1;
    VerifyCodeType type = 2;
}

message SendVerifyCodeResponse {
    common.Result result = 1;
    int32 retry_after = 2;
}

// ==================== 注册 ====================

message RegisterRequest {
    string mobile = 1;
    string verify_code = 2;
    string password = 3;
    string display_name = 4;
}

message RegisterResponse {
    common.Result result = 1;
    UserInfo user = 2;
    TokenPair tokens = 3;
}

// ==================== 登录 ====================

message LoginByPasswordRequest {
    string mobile = 1;
    string password = 2;
}

message LoginByPasswordResponse {
    common.Result result = 1;
    UserInfo user = 2;
    TokenPair tokens = 3;
}

message LoginByCodeRequest {
    string mobile = 1;
    string verify_code = 2;
}

message LoginByCodeResponse {
    common.Result result = 1;
    UserInfo user = 2;
    TokenPair tokens = 3;
}

// ==================== Token 管理 ====================

message RefreshTokenRequest {
    string refresh_token = 1;
}

message RefreshTokenResponse {
    common.Result result = 1;
    TokenPair tokens = 2;
}

message LogoutRequest {
    string refresh_token = 1;
}

message LogoutResponse {
    common.Result result = 1;
}

// ==================== 密码管理 ====================

message ResetPasswordRequest {
    string mobile = 1;
    string verify_code = 2;
    string new_password = 3;
}

message ResetPasswordResponse {
    common.Result result = 1;
}

// ==================== Token 验证（内部服务调用） ====================

message ValidateTokenRequest {
    string access_token = 1;
}

message ValidateTokenResponse {
    common.Result result = 1;
    bool valid = 2;
    string user_id = 3;
    google.protobuf.Timestamp expires_at = 4;
}

// ==================== 服务定义 ====================

service AuthService {
    // 验证码
    rpc SendVerifyCode(SendVerifyCodeRequest) returns (SendVerifyCodeResponse);
    
    // 注册
    rpc Register(RegisterRequest) returns (RegisterResponse);
    
    // 登录
    rpc LoginByPassword(LoginByPasswordRequest) returns (LoginByPasswordResponse);
    rpc LoginByCode(LoginByCodeRequest) returns (LoginByCodeResponse);
    
    // Token 管理
    rpc RefreshToken(RefreshTokenRequest) returns (RefreshTokenResponse);
    rpc Logout(LogoutRequest) returns (LogoutResponse);
    
    // 密码（用户忘记密码，通过手机验证码来重置密码）
    rpc ResetPassword(ResetPasswordRequest) returns (ResetPasswordResponse);
    
    // 内部验证
    rpc ValidateToken(ValidateTokenRequest) returns (ValidateTokenResponse);
}
